diff --git a/src/nouveau/compiler/nak/opt_instr_sched_prepass.rs b/src/nouveau/compiler/nak/opt_instr_sched_prepass.rs
index 854f061b017..ed81a3ae000 100644
--- a/src/nouveau/compiler/nak/opt_instr_sched_prepass.rs
+++ b/src/nouveau/compiler/nak/opt_instr_sched_prepass.rs
@@ -488,9 +488,10 @@ impl<'a> GenerateOrder<'a> {
             }
 
             // Pick an instruction to schedule
-            let next_idx = if used_regs[RegFile::GPR]
+            let next_idx = if (used_regs[RegFile::GPR]
                 <= thresholds.heuristic_threshold
-                && used_regs[RegFile::Pred] < self.max_regs[RegFile::Pred]
+                && used_regs[RegFile::Pred] < self.max_regs[RegFile::Pred])
+                || used_regs[RegFile::Carry] > 0
             {
                 let ReadyInstr { index, .. } = ready_instrs.pop_last().unwrap();
                 index
